### PR TITLE
Add streaming replies for dynamic engine

### PR DESCRIPTION
**Pull request by nvcsathe**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Adds streaming support (delta of newly generated tokens) for the dynamic engine+openai frontend.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/engines/dynamic_engine.py [hunk 1]
```diff
@@ -2112,6 +2114,34 @@ async def async_bookkeep(
...
+                self._partial_emit_lengths.pop(record.requests[-1].request_id, None)
+            partials: list = []
+            for rid, entry in self.requests.items():
+                request = entry.record[-1]
+                if not getattr(request.sampling_params, "streaming", False):
+                    continue
+                already = self._partial_emit_lengths.get(rid, 0)
+                total = len(request.generated_tokens)
+                if total > already:
+                    new_tokens = list(request.generated_tokens[already:])
```
Comment (santhnm2) [inline megatron/core/inference/engines/dynamic_engine.py]: Should this be limited to a maximum number of tokens emitted per partial response? Or is it fine to just send everything that's available?

File: megatron/core/inference/engines/dynamic_engine.py [hunk 2]
```diff
@@ -2434,6 +2464,21 @@ def schedule_requests(self) -> int:
...
+                    # Force active requests to finish on the next step.
+                    request.sampling_params.num_tokens_to_generate = len(request.generated_tokens)
+                    active_ids = self.context.request_ids[: self.context.total_request_count]
+                    matches = torch.where(active_ids == request_id)[0]
+                    if matches.numel() > 0:
+                        idx = int(matches[0].item())
+                        self.context.request_output_lengths[idx] = (
+                            self.context.request_kv_length_offsets[idx]
+                            + self.context.request_query_lengths[idx]
+                        )
```
Comment (santhnm2) [inline megatron/core/inference/engines/dynamic_engine.py:2676]: Why is this accounting necessary here?

File: megatron/core/inference/engines/dynamic_engine.py [hunk 3]
```diff
@@ -2434,6 +2464,21 @@ def schedule_requests(self) -> int:
...
+            elif header == Headers.ABORT_REQUEST:
+                request_id = int(data[1])
+                entry = self.requests.get(request_id)
+                if entry is not None:
+                    request = entry.record[-1]
+                    # Force active requests to finish on the next step.
+                    request.sampling_params.num_tokens_to_generate = len(request.generated_tokens)
+                    active_ids = self.context.request_ids[: self.context.total_request_count]
+                    matches = torch.where(active_ids == request_id)[0]
+                    if matches.numel() > 0:
```
Comment (santhnm2) [inline megatron/core/inference/engines/dynamic_engine.py:2670]: Maybe add an assert within this if condition that `matches.numel() == 1`

Comment (nvcsathe) [reply megatron/core/inference/engines/dynamic_engine.py:2670]: Yeah that should be the only other valid option. I can add that.

File: megatron/core/inference/engines/dynamic_engine.py (hunk 2 above)
Comment (nvcsathe) [reply megatron/core/inference/engines/dynamic_engine.py:2676]: If a request is already in context, this just resets the max seq length for the request to the already generated length so it finishes and is removed in the next iteration

File: megatron/core/inference/engines/dynamic_engine.py (hunk 1 above)
Comment (nvcsathe) [reply megatron/core/inference/engines/dynamic_engine.py]: This would emit one token per decode step or 2-3 in case of MTP because it returns after every engine loop

Comment (santhnm2) [reply megatron/core/inference/engines/dynamic_engine.py]: Ah I vaguely remember from vLLM streaming that they buffer some number of tokens before emitting a partial response - is that still the case?

Comment (nvcsathe) [reply megatron/core/inference/engines/dynamic_engine.py]: Yeah I think vllm does that if it has tokens still pending to be sent in the queue. Then newer tokens are just appended to whatever is in the queue. Here, each token is already sent in a ZMQ send operation so new tokens cant be appended directly.

File: megatron/core/inference/data_parallel_inference_coordinator/handlers.py [hunk 4]
```diff
@@ -192,7 +194,10 @@ def handle_engine_reply(coordinator, sender_identity, payload):
         client_identity = coordinator.request_id_to_client_id[fid]
         client_request_identity = coordinator.request_id_to_client_request_id[fid]
         del coordinator.request_id_to_client_id[fid]
-        del coordinator.request_id_to_client_request_id[fid]
+        original_request_id = coordinator.request_id_to_client_request_id.pop(fid)
```
Comment (shanmugamr1992) [inline megatron/core/inference/data_parallel_inference_coordinator/handlers.py]: NIT  : client_request_id instead of original_request_id ?

**Comment by shanmugamr1992** [comment]
@nvcsathe can you add an example script in https://github.com/NVIDIA/Megatron-LM/tree/main/examples/inference so that users know how to use this. (I will also pull this into the user guide when that gets merged) . 
Also just wanted to know how speculative tokens which are discarded are handled. Apart from that everything else looks good.

File: megatron/core/inference/engines/dynamic_engine.py (hunk 1 above)
Comment (nvcsathe) [reply megatron/core/inference/engines/dynamic_engine.py]: I'll modify this behavior to be similar to vllm because the per token reply increases latency

File: megatron/core/inference/engines/dynamic_engine.py [hunk 5]
```diff
@@ -2007,6 +2009,48 @@ async def async_forward(self) -> Tuple[Dict, Dict, float]:
...
+
+        if not partials:
+            return
+
+        payload = msgpack.packb(
+            [Headers.ENGINE_REPLY_PARTIAL.value, partials], use_bin_type=True
+        )
+        nvtx_range_push("coordinator_streaming")
+        try:
+            self.socket_for_receiving_requests.send(payload, flags=zmq.NOBLOCK)
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/engines/dynamic_engine.py]: We use simple blocking sends w/ this socket - https://github.com/nvcsathe/Megatron-LM/blob/4c8a0c76536293ba2dff8eee64e84b6adf1e20d8/megatron/core/inference/engines/dynamic_engine.py#L968.

Even with a blocking send, the actual transmission is still asynchronous. The runtime copies your payload into ZMQ's internal buffers and returns to you.

**Review by sidsingh-nvidia — CHANGES_REQUESTED** [review]

**Review by santhnm2 — APPROVED** [review]

**Review by sidsingh-nvidia — APPROVED** [review]
Will test the tool-calling version on swe-bench.

File: megatron/core/inference/data_parallel_inference_coordinator/handlers.py [hunk 6]
```diff
@@ -217,6 +222,47 @@ def handle_engine_reply(coordinator, sender_identity, payload):
...
 
 
+@message_handler(Headers.ENGINE_REPLY_PARTIAL)
+def handle_engine_reply_partial(coordinator, sender_identity, payload):
+    """Route incremental engine replies without releasing request routing state."""
+    assert sender_identity in coordinator.identities_of_data_parallel_ranks
+    for partial in payload[1]:
+        request_id = partial["request_id"]
+        client_identity = coordinator.request_id_to_client_id.get(request_id)
+        if client_identity is None:
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/data_parallel_inference_coordinator/handlers.py]: Shouldn't we throw an error here?

Comment (sidsingh-nvidia) [reply megatron/core/inference/data_parallel_inference_coordinator/handlers.py]: In what cases could the client identity be None?

File: megatron/core/inference/data_parallel_inference_coordinator/handlers.py [hunk 7]
```diff
@@ -217,6 +222,47 @@ def handle_engine_reply(coordinator, sender_identity, payload):
...
+            continue
+        client_request_id = coordinator.request_id_to_client_request_id[request_id]
+        coordinator.router_socket.send_multipart(
+            [
+                client_identity,
+                msgpack.packb(
+                    [Headers.ENGINE_REPLY_PARTIAL.value, client_request_id, partial],
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/data_parallel_inference_coordinator/handlers.py:241]: Why are we not detokenizing the partial replies?

File: megatron/core/inference/data_parallel_inference_coordinator/handlers.py [hunk 8]
```diff
@@ -87,6 +87,7 @@ def handle_submit_request(coordinator, sender_identity, payload):
     coordinator.next_request_id += 1
     coordinator.request_id_to_client_id[request_id] = sender_identity
     coordinator.request_id_to_client_request_id[request_id] = client_request_id
+    coordinator.client_request_to_request_id[(sender_identity, client_request_id)] = request_id
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/data_parallel_inference_coordinator/handlers.py:90]: Can this be a map from client_request_id to request_id? Then we can do a join with request_id_to_client_id to get the sender_identity.

File: megatron/core/inference/data_parallel_inference_coordinator/handlers.py [hunk 9]
```diff
@@ -117,6 +118,7 @@ def handle_submit_request(coordinator, sender_identity, payload):
         logging.error("Coordinator: no reachable engines for request %d", request_id)
         del coordinator.request_id_to_client_id[request_id]
         del coordinator.request_id_to_client_request_id[request_id]
+        coordinator.client_request_to_request_id.pop((sender_identity, client_request_id), None)
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/data_parallel_inference_coordinator/handlers.py]: can we use del here for consistency? pop with None might silently ignore errors.

File: megatron/core/inference/data_parallel_inference_coordinator/handlers.py [hunk 10]
```diff
@@ -198,7 +200,10 @@ def handle_engine_reply(coordinator, sender_identity, payload):
         client_identity = coordinator.request_id_to_client_id[fid]
         client_request_identity = coordinator.request_id_to_client_request_id[fid]
         del coordinator.request_id_to_client_id[fid]
-        del coordinator.request_id_to_client_request_id[fid]
+        client_request_id = coordinator.request_id_to_client_request_id.pop(fid)
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/data_parallel_inference_coordinator/handlers.py]: use del.

**Review by sidsingh-nvidia — CHANGES_REQUESTED** [review]

File: megatron/core/inference/data_parallel_inference_coordinator/handlers.py (hunk 8 above)
Comment (nvcsathe) [reply megatron/core/inference/data_parallel_inference_coordinator/handlers.py:90]: Here every client_request_id starts at 0 for every InferenceClient, so (sender_identity, client_request_id) is the unique key for a lookup

Comment (sidsingh-nvidia) [reply megatron/core/inference/data_parallel_inference_coordinator/handlers.py:90]: Makes sense

File: megatron/core/inference/data_parallel_inference_coordinator/handlers.py (hunk 6 above)
Comment (nvcsathe) [reply megatron/core/inference/data_parallel_inference_coordinator/handlers.py]: Yes, this should be an error

File: megatron/core/inference/data_parallel_inference_coordinator/handlers.py (hunk 7 above)
Comment (nvcsathe) [reply megatron/core/inference/data_parallel_inference_coordinator/handlers.py:241]: Because detokenizing tokens independently might produce the wrong text. The frontend handles incremental detokenization

Comment (sidsingh-nvidia) [reply megatron/core/inference/data_parallel_inference_coordinator/handlers.py:241]: Can we put it as a comment here?

**Review by sidsingh-nvidia — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]
